Update Readme with better structure and missing information (#144)
lblod/frontend-embeddable-notule-editor: README.md

## Editor API
The RDFa editor uses [the Prosemirror toolkit](https://prosemirror.net/) as a base. After the `editorElement.initEditor()` function is called you will have access to the editor methods, including the controller with `editorElement.controller`. This controller is an instance of the [SayController](https://github.com/lblod/ember-rdfa-editor/blob/d4472d2e237256d30333cfcc20ce6eea7db241f2/addon/core/say-controller.ts) class of the [ember-rdfa-editor](https://github.com/lblod/ember-rdfa-editor). 

#### editorElement API
These are function available from the editor element, which is the HTML element with the class `notule-editor`. 
- `initEditor(arrayOfPluginNames: string[], userConfigObject)`: Initialize the editor by passing an array of plugin names that should be activated and an object that contains the configuration for the editor and its plugins. See [Managing Plugins](managing-plugins) for more info. 
  :warning: **This has to be called before accessing any other methods**.
- `enableEnvironmentBanner()`: enable the banner that shows the environment and versions of plugins used.
- `disableEnvironmentBanner()`: disable the banner.

> - `controller`: direct access the the [SayController](https://github.com/lblod/ember-rdfa-editor/blob/d4472d2e237256d30333cfcc20ce6eea7db241f2/addon/core/say-controller.ts) object. See [controller API](controller-api).

Reviewer: ```suggestion
- `controller`: provides direct access to a [SayController](https://github.com/lblod/ember-rdfa-editor/blob/d4472d2e237256d30333cfcc20ce6eea7db241f2/addon/core/say-controller.ts) object. See [controller API](controller-api).
```

- `setHtmlContent(content: string)`: set the HTML content inside the editor, overwriting all previous content.  

> - `getHtmlContent()`: Get the HTML content of the editor. This can be different than custom content set via `setHtmlContent`, because of HTML parsing logic.

Reviewer: ```suggestion
- `getHtmlContent()`: Get the HTML content of the editor. This might be different than custom content set via `setHtmlContent`, because of HTML parsing logic.
```

- `withTransaction(callback: (tr: Transaction) => Transaction | null, includeEmbeddedView = false)`: apply a [Prosemirror transaction](https://prosemirror.net/docs/ref/#state.Transaction) on the main view (or currently active embedded view). When you want to apply the transaction, the callback should return the transaction object.
- `mainEditorState`: the [editor state](https://prosemirror.net/docs/ref/#state.Editor_State) instance of the main editor
- `activeEditorState`: the state instance of the active editor. This is usually the same as the `mainEditorState`, except when inside a nested instance (like in a variable field). For typical use, the `mainEditorState` is the one you will most likely need.
- `mainEditorView`: the [view](https://prosemirror.net/docs/ref/#view.EditorView) instance of the main editor
- `activeEditorView`: the view instance of the active editor. (see above for the distinction between the main and active editor)

> - `setActiveView(view: RdfaEditorView)`: activate a specific view.

Reviewer: I'm not sure if this method needs to be documented for embeddable, as it has quite a specific use case for embedded editors.

Author: I think the reason is that if they want to use custom commands, it might work differently/weirdly for variables and such. But I am also not sure if there are many cases where a command they'll write will really need to know the difference.
Maybe leaving it out but instead add a note at the bottom about it is a better idea?

Reviewer: Yes, maybe or in a section on custom prosemirror commands.

Author: @elpoelma I added a separate section with some explanation about prosemirror commands, linking to the editor wiki.
The explanation of what the difference is between the main and active editor state should really be part of that wiki instead of trying to explain it here without (the wiki has context to make the explanation easier).

If you could check the new addition if it is okay, that would be great :)

Reviewer: looks great!

# Configuring The Editor

The editor can be customized to best fit your application. 
* [Managing Plugins](#managing-plugins): A list of plugins you can enable, including explanation of how to use them
* [Environment banner](#enabling-disabling-the-environment-banner): how to enable/disable this banner
* [Localization](#localization): language options in the editor
* [styling](#styling)

## Managing Plugins
Embeddable ships with the following plugins available. 

> This Readme contains all the important info and configuration for the plugins. For more technical and Ember-specific explanation of every plugin, you can check out the Readme of [lblod/ember-rdfa-editor-lblod-plugins](https://github.com/lblod/ember-rdfa-editor-lblod-plugins).

Reviewer: ```suggestion
This Readme contains all the important info and configuration for the plugins. For more technical and Ember-specific explanations of every plugin, you can check out the Readme of [lblod/ember-rdfa-editor-lblod-plugins](https://github.com/lblod/ember-rdfa-editor-lblod-plugins).
```

> Every plugin can be enabled by passing its name to `arrayOfPluginNames` array and its configuration to `userConfigObject` with the initialization function `initEditor(arrayOfPluginNames, userConfigObject)`.

Reviewer: ```suggestion
Every plugin can be enabled by passing its name to the `arrayOfPluginNames` array and its configuration to `userConfigObject` with the initialization function `initEditor(arrayOfPluginNames, userConfigObject)`.
```

> :warning: The values shown for the config are default values used if you do not pass a config. If you pass **any** config for a plugin, it will not use any of these defaults and only use the config provided. Make sure to pass all required attributes in this case, even if you do not change them.

* [article-structure](#article-structure): Provides structures to better manage official documents, like titles, chapters, articles and paragraphs. It allows you to insert, move and delete them. It has two modes: `besluit` for besluit articles and `regulatoryStatement` for all other structures.

> * [besluit](#besluit): Provides the correct rdfa-structure for constructing a decisions ("besluiten") and some basic validation if mandatory structures are present in the document.

Reviewer: ```suggestion
* [besluit](#besluit): Provides the correct rdfa-structure for constructing a decision ("besluiten") and some basic validation which checks whether mandatory structures are present in the document.
```

* [citation](#citation): Search and insert references to citations (a legal resource/expression)

> * [rdfa-date](#rdfa-date): Inserting and modify annotated date and times

Reviewer: ```suggestion
* [rdfa-date](#rdfa-date): Inserting and modifying annotated dates and times
```

> * [roadsign-regulation](#roadsign-regulation): Insert roadsign regulation, based on the registry managed and provided by MOW (Mobiliteit en Openbare Werken)

Reviewer: ```suggestion
* [roadsign-regulation](#roadsign-regulation): Insert a roadsign regulation, based on the registry managed and provided by MOW (Mobiliteit en Openbare Werken)
```

* [table-of-contents](#table-of-contents): Show a table of contents with clickable sections defined by [article-structure](#article-structure)
* [variable](#rdfa-variables): Allows insertion and filling in of custom rdfa variables
* [formatting-toggle](#formatting-toggle): Allows to toggle the formatting marks with a button
* [rdfa-blocks-toggle](#rdfa-blocks-toggle): Allows to toggle the visual indications of the rdfa blocks with a button.
* [template-comments](template-comments): Allows insertion and editing of comment blocks to provide extra information to a user filling in a document. These are visually distinct units with a special RDFa type, which allows them to be filtered out during postprocessing.
##### General Config options
There are some options you can pass to `pluginsConfig` in `initEditor` that are not connected to a plugin.
- `docContent: 'block+'`: The property docContent specifies which nodes are allowed in the document. By default we allow one or more nodes of the supertype block, which includes most content. For more info about this check the [Prosemirror docs](https://prosemirror.net/docs/guide/#schema.content_expressions). 

>   See `public/test.html` where `docContent` is specified to allow [article-structure](#Article Structure) nodes in a specific order.

Reviewer: ```suggestion
  See `public/test.html` where `docContent` is specified to allow [article-structure](#article-structure) nodes in a specific order.
```

### Article Structure
This plugin is in charge of inserting and manipulating structures. There are insertion buttons in the insert menu of the right sidebar. The plugin has two modes, being `besluit` and `regulatoryStatement`. 

- `besluit` Mode
	- Only allowed to insert articles
- `regulatoryStatement` mode
	- Able to insert titles, chapters, sections... 
	- [table-of-contents](#Table of Contents) default config works with this mode.

> After inserting a structure and selecting it, a card will show to move and delete the structure. These might be disabled if the action is not possible. 

Reviewer: ```suggestion
After inserting a structure and selecting it, a card will show options to move and delete the structure. These might be disabled if the action is not possible. 
```

Using the button "with content" will also delete everything included in the structure, instead of just the closest heading. 
![article structure card](https://imgur.com/2zkbNw3.png)
***
:heavy_plus_sign: Enable by adding `"article-structure"` to `arrayOfPluginNames` array.

```javascript
// pass to pluginsConfig
articleStructure: {
  mode: 'besluit',
}
```
The options for mode are:
- 'besluit' (default): for manipulating articles in decisions. 
- 'regulatoryStatement': for manipulating chapters, sections, etc as regulatory statements.

### Besluit 
 :warning: The besluit plugin is incompatible with the `regulatoryStatement` mode of  [article-structure](#Article Structure).

This will add needed rdfa structures to create a besluit and add some validation. There is no direct interaction with the plugin. The validation will pop up in the sidebar. You can try this out by deleting the title and seeing the following pop up:
![besluit plugin](https://imgur.com/iwCudJy.png)
***

> :heavy_plus_sign: Enable by adding `"besluit"` to `arrayOfPluginNames` array.

Reviewer: ```suggestion
:heavy_plus_sign: Enable by adding `"besluit"` to the `arrayOfPluginNames` array.
```

No configuration is needed.

### Citation
Add the possibility to add references to specific legal documents. There are two ways to use this plugin

**A. Insert Button**

> Click `insert reference` button in the right sidebar

Reviewer: ```suggestion
Click the `insert reference` button in the right sidebar.
```

> This will open a modal where you can search for different type of legal documents, preview them and insert if desired.

Reviewer: ```suggestion
This will open a modal where you can search for different types of legal documents, preview them and insert them if desired.
```

**B. Type Keyword**
Type one of the trigger phrases, where `[words to search for]` will be filled in as a search term.
* [specification]**decreet** [words to search for] *(e.g. "gemeentedecreet wijziging")*
* **omzendbrief** [words to search for]
* **verdrag** [words to search for]
* **grondwetswijziging** [words to search for]
* **samenwerkingsakkoord** [words to search for]
* [specification]**wetboek** [words to search for]
* **protocol** [words to search for]
* **besluit van de vlaamse regering** [words to search for]
* **gecoordineerde wetten** [words to search for]
* [specification]**wet** [words to search for] *(e.g. "kieswet wijziging", or "grondwet")*
* **koninklijk besluit** [words to search for]
* **ministerieel besluit** [words to search for]
* **genummerd besluit** [words to search for]

After typing this trigger phrase, a card will shop up in the right sidebar with the type and search term filled in. Press `Advanced search` to pop open the same modal as shown in **A.** 
![citation plugin](https://imgur.com/oerd9rV.png)
***

> :heavy_plus_sign: Enable by adding `"citation"` to `arrayOfPluginNames` array.

Reviewer: ```suggestion
:heavy_plus_sign: Enable by adding `"citation"` to the `arrayOfPluginNames` array.
```

citation: {
  endpoint: '/codex/sparql',
  type: 'nodes',
  activeInNodeTypes(schema, _state) {
    // the root node of the document is the doc.
    return new Set([schema.nodes.doc]);
  }
},
```
- `endpoint`: where to fetch the citation data from (the codex)
- `type`: this is `'nodes'` or `'ranges'` and specifies the type of check that can be specified. 
	- if type `'nodes'`:
		- `activeInNodeTypes`: given the Prosemirror schema and editor state, return a `Set` of nodetypes inside which the plugin should be active. Embeddable does not expose the schema directly, so some internal knowledge is needed to use this effectively.
	- if type `'ranges'`: 
		- `activeInRanges`: given the Prosemirror editor state, return an array of ranges for the plugin to be active in, for example `[[0,50], [70,100]]`

Both examples show how to activate the plugin for the *whole* document.

### RDFa Date
A simple plugin to insert and modify *semantic* dates and timestamps in an editor document. When added, an `insert date` button will be available in the right insert sidebar. When selecting a date, a card will show up at the same place to edit this date and choose a format for display. 
***

> :heavy_plus_sign: Enable by adding `"rdfa-date"` to `arrayOfPluginNames` array.

Reviewer: ```suggestion
:heavy_plus_sign: Enable by adding `"rdfa-date"` to the `arrayOfPluginNames` array.
```

  allowCustomFormat: true,
},
```
- `placeholder`:
	- `insertDate`: placeholder to show when inserting a date that has no data yet.
	- `insertDateTime`: placeholder to show when inserting a datetime that has no data yet.
- `formats`: specify default formats to show for selection in the date card.

> 	- `label` (optional): The label shown to the user on the card. If not provided, the format is used instead e.g.: `dd/MMvariable/yyyy`

Reviewer: What is `variable` doing here?

Author: middle-click pasting accident :(

	- `key`: A **unique** identifier used for identification in the internal code. 
	- `dateFormat`: The date format used when this is selected.

> 	- `dateTimeFormat`: The datetime format to use when this is selected. Used when the user selected "Include time".

Reviewer: ```suggestion
	- `dateTimeFormat`: The datetime format to use when this is selected. Used when the user selects "Include time".
```

- `allowCustomFormat`: true/false, determines if the option to insert a fully custom format is available.

For more information about the syntax of defining date(time) formats check the documentation of the underlying library used [date-fns](https://date-fns.org/v2.29.3/docs/format).

### Roadsign Regulation
Add annnotated *mobiliteitsmaatregelen* from a specified registry, which will most likely be using the [public facing sparql endpoint](https://register.mobiliteit.vlaanderen.be/sparql) of [the roadsign registry](https://register.mobiliteit.vlaanderen.be). This data is maintained by experts at [MOW Vlaanderen](https://www.vlaanderen.be/departement-mobiliteit-en-openbare-werken).

:warning: This plugin will only activate in *besluiten* with a certain rdf type.

> <details><summary>Exhaustive list of types</summary>

Reviewer: ```suggestion
<details><summary>Exhaustive list of decision types in which this plugin will activate</summary>
```

> `https://data.vlaanderen.be/id/concept/BesluitType/4d8f678a-6fa4-4d5f-a2a1-80974e43bf34`

Reviewer: Maybe the types in this list should be converted to links?

Author: Right, I didn't think the links would actually go to something (as they don't always...)
but in this case they do, so links make more sense.

`https://data.vlaanderen.be/id/concept/BesluitType/7d95fd2e-3cc9-4a4c-a58e-0fbc408c2f9b`

`https://data.vlaanderen.be/id/concept/BesluitType/3bba9f10-faff-49a6-acaa-85af7f2199a3`

`https://data.vlaanderen.be/id/concept/BesluitType/0d1278af-b69e-4152-a418-ec5cfd1c7d0b`

`https://data.vlaanderen.be/id/concept/BesluitType/e8afe7c5-9640-4db8-8f74-3f023bec3241`

`https://data.vlaanderen.be/id/concept/BesluitType/256bd04a-b74b-4f2a-8f5d-14dda4765af9`

`https://data.vlaanderen.be/id/concept/BesluitType/67378dd0-5413-474b-8996-d992ef81637a`
</details>

When the cursor is inside such a *besluit*, the button `Voeg mobiliteitsmaatregel in` will appear under the insert menu. Clicking this will show a modal to filter and select roadsign regulation to insert.

![roadsign regulation modal](https://i.imgur.com/z7My8lm.png)
***

> :heavy_plus_sign: Enable by adding `"roadsign-regulation"` to `arrayOfPluginNames` array.

Reviewer: ```suggestion
:heavy_plus_sign: Enable by adding `"roadsign-regulation"` to the `arrayOfPluginNames` array.
```

```javascript

// pass to pluginsConfig
roadsignRegulation: {
  endpoint: 'https://dev.roadsigns.lblod.info/sparql',
  imageBaseUrl: 'https://register.mobiliteit.vlaanderen.be/',
}
```
- `endpoint`: The sparql endpoint to fetch roadsigns. By default the development endpoint is used, so make sure to change this in production to your own registry or `https://register.mobiliteit.vlaanderen.be/sparql`.
- `imageBaseUrl`: In production, some old roadsigns of MOW miss a base URL for images, which will be prepend with this URL. If you provide your own registry with correct data, this will not be used.

### Table of Contents
Add a table of contents at the top of the document. It can be toggled with a button in the top toolbar. 

At this time it will only work well together with [article-structure plugin](#article-structure) in `regulatoryStatement` mode by using the default config.

> :warning: For use in different situations, open an issue on this repo with the usecase, so we can help. The [Prosemirror schema](https://prosemirror.net/docs/guide/#schema) that is used in the config is public-facing yet, so changing this is not trivial.

Reviewer: ```suggestion
:warning: For use in different situations, open an issue on this repo with the usecase, so we can help. The [Prosemirror schema](https://prosemirror.net/docs/guide/#schema) that is used in the config is not public-facing yet, so changing this is not trivial.
```

***

> :heavy_plus_sign: Enable by adding `"table-of-contents"` to `arrayOfPluginNames` array.

Reviewer: ```suggestion
:heavy_plus_sign: Enable by adding `"table-of-contents"` to the `arrayOfPluginNames` array.
```

A variable can be inserted with the card shown in the right sidebar.
![insert variable card](https://imgur.com/9kSqgXc.png)

**Types of variables:**
- *text*: a variable that any text can be typed in
- *number*: pops up an input box that will validate constraints and includes a button to show the number in words. Constraints (min/max) can be set when inserting the variable.
- *date*: works just like the [RDFa date plugin](rdfa-date), where a user can input a date in specified formats.
- *location*: choose out of a list of location options, that can contain placeholders themselves. 
- *codelist*: when inserting, a specific codelist has to be chosen. This codelist is a list of values the user can choose from to fill in the variable. Either the user can select one (single selection) or multiple (multiple selection). 
- *address*: when inserted, the user can click this to get a modal for searching addresses from the Belgium address register. This can be used to insert existing addresses.
  **note**: when searching for submunicipalities, only the main municipality will show up in the search. However, when searching for a street, the correct zip-code will be used.
***

> :heavy_plus_sign: Enable by adding `"variable"` to `arrayOfPluginNames` array.

Reviewer: ```suggestion
:heavy_plus_sign: Enable by adding `"variable"` to the `arrayOfPluginNames` array.
```

	- `locationEndpoint`: the endpoint to fetch location options, which will be added to the location variable's RDFa and used as the endpoint when selecting a location variable. For production you'll likely want to use `https://register.mobiliteit.vlaanderen.be`.
- `edit`: configuration for editing an inserted variable
	- `enable`: is editing a variable allowed (removing is always possible!)
	- `location`: config for location variable
		- `endpoint`: *fallback* endpoint for location variable if the variable is missing the endpoint in its RDFa. This will most likely be the same as the endpoint used for inserting.
		- `zonalLocationCodelistUri`: the URI to search for if the location variable is included in a zonal traffic measure.
		- `nonZonalLocationCodelistUri`: the URI to search for if the location variable is included in a non-zonal traffic measure.

### Formatting Toggle
This will add a button in the top toolbar `Show formatting marks`. This toggles the visibility of all formatting marks of the document such as break lines, paragraphs...
![document with formatting annotations](https://imgur.com/KTNxuBW.png)
***

> :heavy_plus_sign: Enable by adding `"formatting-toggle"` to `arrayOfPluginNames` array.

Reviewer: ```suggestion
:heavy_plus_sign: Enable by adding `"formatting-toggle"` to the `arrayOfPluginNames` array.
```

No configuration is needed.

### Rdfa Blocks Toggle
This will add a button in the top toolbar `Show annotations`. This toggles the visibility of RDFa information contained in the document. This is useful if you want to check for errors in the RDFa structure, or simply have a look at what data the editor is generating behind the scenes. As such, it is mostly useful for expert users.

![document with rdfa blocks visible](https://imgur.com/Asdu2aN.png)
***

> :heavy_plus_sign: Enable by adding `"rdfa-blocks-toggle"` to `arrayOfPluginNames` array.

Reviewer: ```suggestion
:heavy_plus_sign: Enable by adding `"rdfa-blocks-toggle"` to the `arrayOfPluginNames` array.
```

## Installation

* `git clone https://github.com/lblod/frontend-embeddable-notule-editor.git` this repository
* `cd frontend-embeddable-notule-editor`
* `npm install`

## Running
* `ember serve`
* Visit an example of the embeddable at [http://localhost:4200/test.html](http://localhost:4200/test.html).
### Linting
* `npm run lint:hbs`
* `npm run lint:js`
* `npm run lint:js -- --fix`

### Building
* `ember build` (development)
* `ember build --environment production` (production)
### Releasing
Take special care when releasing a new version of this.
- The changelog is automatically created using the PRs. 
- To avoid a connection to Github, build files are not automatically added to the release

> 	- Add the files under `dist/assets` from a production build to the release in a .zip archive. 

Reviewer: I think it should be all files under `dist`?

Author: I'm actually not sure. Are the other files referenced in the assets files, and that's the reason they are needed?
It is not really specified in the readme (now or before) if you should also "host" those in the same place, or if only the ones under `assets` are needed and the rest can be ignored.

Either way (if they are needed or not needed), it should be more explicit, so definitely needs an update around that.

Reviewer: In the previous releases, we always bundled the whole `dist` folder I think. But I don't know if the files outside of `assets` are needed.

Author: I did a small test and as far as I can tell they are not needed, everything seems to works as expected. I might be missing something though :shrug:

Reviewer: ok :) makes sense!

Author: added a bit more explicit explanation, so there is more consistency between the different ways of using Embeddable.

### Developing
#### How it works
This repository includes [the editor](https://github.com/lblod/ember-rdfa-editor) and [editor plugins](https://github.com/lblod/ember-rdfa-editor-lblod-plugins) packaged together with Ember so the editor can be used in projects outside of Ember. This is mostly done by adjust the build process in `ember-cli-build.js`, by specifying the output filename and limiting the chunks to one file, for easy importing.

> the consumer loads the editor in their own div element. This editor is fully defined in `app/components/simple-editor.js`, with consumer-facing logic bound in `insertedInDom` and  logic that needs a controller bound in `handleRdfaEditorInit`. 

Reviewer: ```suggestion
The consumer loads the editor in their own div element. This editor is fully defined in `app/components/simple-editor.js`, with consumer-facing logic bound in `insertedInDom` and  logic that needs a controller bound in `handleRdfaEditorInit`. 
```
